packages/batch/tests/unit/BatchProcessor.test.ts

```diff
@@ -0,0 +1,477 @@
+/**
+ * Test BatchProcessor class
+ *
+ * @group unit/batch/class/batchprocessor
+ */
+import { BatchProcessingError, BatchProcessor, EventType } from '../../src';
+import { DynamoDBRecord, KinesisStreamRecord, SQSRecord } from 'aws-lambda';
+import {
+  sqsEventFactory,
+  kinesisEventFactory,
+  dynamodbEventFactory,
+} from '../../tests/helpers/factories';
+import {
+  sqsRecordHandler,
+  asyncSqsRecordHandler,
+  kinesisRecordHandler,
+  asyncKinesisRecordHandler,
+  dynamodbRecordHandler,
+  asyncDynamodbRecordHandler,
+} from '../../tests/helpers/handlers';
+
+describe('Class: BatchProcessor', () => {
+  const ENVIRONMENT_VARIABLES = process.env;
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+    jest.resetModules();
+    process.env = { ...ENVIRONMENT_VARIABLES };
+  });
+
+  afterAll(() => {
+    process.env = ENVIRONMENT_VARIABLES;
+  });
+
+  describe('Synchronously processing SQS Records', () => {
+    test('Batch processing SQS records with no failures', async () => {
+      // Prepare
+      const firstRecord = sqsEventFactory('success');
+      const secondRecord = sqsEventFactory('success');
+      const records = [firstRecord, secondRecord];
+      const processor = new BatchProcessor(EventType.SQS);
+
+      // Act
+      processor.register(records, sqsRecordHandler);
+      const processedMessages = await processor.process();
```

Comment:
If `sqsRecordHandler` is synchronous, do we need `await`?

Reply:
The `process()` function is asynchronous so that it can handle the asynchronous functions, I'm not sure if there's an alternative for that.

Reply:
Let's take these two example handlers:
```ts
const mySyncHandler = (record) => {
  // do something
};

const myAsyncHandler = async (record) => {
  // do something async
};
```

Depending on which one of the two I pass to `processor.register()` we should have to `await` the `processor.process()` function or not.

For instance:
```ts
processor.register(records, mySyncHandler);
const processedMessages = processor.process(); // no await, since mySyncHandler is synchronous

processor.register(records, myAsyncHandler);
const processedMessagesAsync = await processor.process() // using await since myAsyncHandler is async (aka it returns a promise)
```

We should aim at supporting this type of behavior, which might require to be creative not only with the implementation but also with the return types (i.e. return type of `process()` method depend on type of the second arg of `register()` method).

I'm pointing this out at this stage to set direction, however if we don't have a solution right away I am okay with putting a pin on it and opening an issue. This way we can address this later on.

Reply:
I'll have a look today to see if I can figure anything out and aim for a revision by EOD for me for you to be able to review tomorrow. I'll leave a comment here if I manage to implement something that works.

Reply:
From what I've looked at, it seems like our options are either to create a separate set of functions (like what Python has) or to always await (what we have now).

Theoretically, if we wanted to keep it as a single method, we might be able to do something like [this](https://stackoverflow.com/questions/72202437/refactor-both-sync-and-async-method-implementations), but we'd have to create another level on top of `process()` that returns a list or a promised list. Then, we'd be able to have an additional `AsyncBatchProcessor` with an async `process()` implementation that generates that list with a `Promise` wrapper properly.

Reply:
As discussed during our meeting for now we will leave the implementation as-is and we will revisit the topic once the utility is in a more advanced state. I'll open an issue and link it here shortly.

Reply:
Issue created #1603, resolving the conversation.